crates/gen/src/class.rs

```diff
@@ -161,31 +160,37 @@ impl Class {
 
             quote! {
                 #[repr(transparent)]
-                #[derive(::std::clone::Clone, ::std::default::Default, ::std::cmp::PartialEq)]
-                pub struct #name { ptr: ::winrt::ComPtr<#default_name> }
+                pub struct #name(::winrt::Object);
                 impl #name {
                     #new
                     #methods
                     #async_get
                     #call_factory
                 }
+                impl ::std::clone::Clone for #name {
+                    fn clone(&self) -> Self {
+                        Self(self.0.clone())
+                    }
+                }
+                impl ::std::cmp::PartialEq for #name {
```

Comment:
As I read it, [`PartialEq`](https://doc.rust-lang.org/std/cmp/trait.PartialEq.html) (as well as [`Eq`](https://doc.rust-lang.org/std/cmp/trait.Eq.html)) represent *equality* while this implementation models an *identity* relation. This has two consequences:

* Two references to unrelated object instances representing the exact same information would compare as unequal. Though perfectly justifiable this feels as awkward as it is surprising.
* Two references to the same object instance always compare as equal, even when a piecewise comparison over the object's state data wouldn't. This can happen for types exposing a floating point property with a `NaN` value. While rare, this causes a disconnect between object comparison and constituent object state comparison.

Much as I understand that mixing value semantics with a (mostly) reference-based system isn't going to be as smooth as baby skin, is this something that clients of winrt-rs need to be aware of?

Reply:
This is a very good point. We actually might want to remove `PartialEq` and `Eq` support for these types and add a bespoke `eq` method on them that allows for this kind of equality testing. You are correct that in Rust `==` strictly meaning structural equality and not reference equality.

Reply:
In practice, I don't think this matters too much. I also don't see a strict requirement in Rust for equality to equate to structure. The [first example I came across](https://doc.rust-lang.org/std/collections/index.html#insert-and-complex-keys) does not meet this requirement. You simply cannot compare the structure of two objects in COM or WinRT because their implementations are hidden nor is it necessarily meaningful. 

The main reason WinRT objects implement equality and ordering operators or traits in various languages (beyond identity comparisons) is to support ordered containers. We had the same debates about C# and C++ with respect to equality and ordering and arrived at the same conclusion that the only meaningful and strictly correct comparison is via object identity.

But I do agree that this is something that we should document so that developers are aware of this behavior as it is foundational to all COM-based type systems.